### DISCUSSION TITLE
Scorecard won't plot stat-analysis output due to event equalization

### DISCUSSION THREAD

**Discussion by briannen** [dtcenter/METplus#2734] [Plot Generation]
I have created a scorecard using output from point-stat that works correctly, but has too many lead times to be useful. I used stat-analysis to aggregate the lead times to 24 hours, but now the event equalization is throwing out all of the data and the scorecard won't plot. Why does event equalization affect the ability to plot the 24 hour averages using stat-analysis (vs. output from each forecast time)?

**Comment by JohnHalleyGotway** [comment]
Hi @briannen, I see you're having issues with scorecard generation. I read through your question and see there's no easy, simple answer to it. As is often the case, the answer is in the details.

Can you please provide me more information so that I'm able to replicate the behavior you're seeing?

- On what machine are you running?
- Please provide the scorecard XML and command that successfully generate the the scorecard with "too many lead times to be useful".
- Please provide the scorecard XML and command that discard all the data based based on event equalization.

Once I'm able to replicate the behavior you describe, hopefully I'll be able to offer advice on how to proceed.

**Reply by briannen** [reply]
I'm running on mohawk.

The working scorecard XML is /d2/projects/land_nysm/metv/scorecards/test_scorecard.xml.
The non-working scorecard XML is /d2/projects/land_nysm/metv/scorecards/scorecard_agg_leads.xml.

For both scorecards, I use the mv_scorecard.sh script from /d2/projects/land_nysm/metv. The command is ./mv_scorecard.sh scorecards/{file}.xml

Thanks!

**Reply by JohnHalleyGotway** [reply]
Thanks @briannen for pointing me to your work location on mohawk. I copied your XML files, modified `test_scorecard.xml` to direct the output to my working directory, and ran METviewer.

I DO NOT get a successful run. Instead, I'm seeing error messages about all the data being event-equalized out:
```
> cd /d2/personal/johnhg/METplus/discussions/2734
> ./mv_scorecard.sh test_scorecard.xml | tee test_scorecard.log
> egrep -i "error|equal" test_scorecard.log 
WARNING: event equalization removed 10 rows
INFO: Event equalization has produced no results.  Data frame is empty.
WARNING: event equalization removed 10 rows
INFO: Event equalization has produced no results.  Data frame is empty.
2024-10-18T19:33:02,003 ERROR [-1] [printFormattedTable] e.u.m.s.d.DatabaseManagerSql: 
  **  ERROR: Caught class edu.ucar.metviewer.EmptyResultSetException in printFormattedTable(ResultSet res): result set contained no data
INFO: Event equalization has produced no results.  Data frame is empty.
...
```
I see that your `data` directory does contain output from Sep 30 and Oct 11. I wonder if perhaps that pre-existing data is giving you a false positive, making you think it's working when it isn't?
```
ls -lah /d2/projects/land_nysm/metv/scorecards/data/
-rw-r--r-- 1 briannen rap 2.3K Oct 11 18:50 test_scorecard_agg.agg_stat.info
-rw-r--r-- 1 briannen rap    1 Oct 11 18:50 test_scorecard_agg.data
-rw-r--r-- 1 briannen rap  130 Oct 11 18:50 test_scorecard_agg.data.agg_stat
-rw-r--r-- 1 briannen rap  14K Sep 30 14:24 test_scorecard.agg_stat.info
-rw-r--r-- 1 briannen rap 319K Sep 30 14:25 test_scorecard.data
-rw-r--r-- 1 briannen rap 325K Sep 30 14:24 test_scorecard.data.agg_stat
```
Please try re-testing `test_scorecard.xml` and either redirect the output to a new, empty directory or first empty out the subdirectories of `/d2/projects/land_nysm/metv/scorecards`.

We're assuming the problem lies in the difference between `test_scorecard.xml` and `scorecard_agg_leads.xml` but it might actually lie in `test_scorecard.xml` itself.

**Reply by briannen** [reply]
I realized that I didn't just add the stat-analysis output to the database, but I dropped and reloaded completely. I will have to reload the database with all of the stat-analysis and point-stat output first, but then I will re-test the scorecard xml.

**Reply by briannen** [reply]
Okay, I have re-tested the scorecard xmls. I am now running in /d2/projects/land_nysm/metv/scorecards_troubleshoot on mohawk. As you can see by the attached images, the non-aggregated scorecard (test_scorecard.xml) is working, but the aggregated scorecard (scorecard_agg_leads.xml) is not.
![test_scorecard](https://github.com/user-attachments/assets/507d73fc-31e8-4446-a457-536b0e1784ef)
![test_scorecard_agg](https://github.com/user-attachments/assets/5bc47fe5-edb3-4813-8d72-599c6facabf6)

**Comment by JohnHalleyGotway** [comment]
@briannen thanks for reloading the database on mohawk. I reran using the same XML file you sent to me but am seeing a very different result.

Here's my commands on mohawk:
```
cd /d2/personal/johnhg/METplus/discussions/2734
 ./mv_scorecard.sh scorecard_agg_leads.xml | tee scorecard_agg_leads.log
```
I do note that there are 272 log messages about "discarding" data:
```
grep discarding scorecard_agg_leads.log | wc -l
272
```
But the resulting scorecard image does not match yours at all. I did remove all the "intermediate" files in the `data`, `plots`, and `scripts` directories before running this to avoid "old" data impacting the new results.
![test_scorecard_agg](https://github.com/user-attachments/assets/d8d2df99-e808-424b-b26b-fb158d41b205)
So it still isn't obvious to me yet exactly what needs to be debugged.

I'd recommend that you...
1. Compare `/d2/personal/johnhg/METplus/discussions/2734/scorecard_agg_leads.xml` to the version of the `scorecard_agg_leads.xml` to make sure there aren't obvious diffs (other than paths).
2. Delete the working contents of the `data`, `plots`, and `scripts` directories.
3. Rerun `scorecard_agg_leads.xml` to regenerate `test_scorecard_agg.png`.

I mean on the same machine (mohawk), we should be able run the same command with the same inputs to get exactly the same result. Are you still getting all blue non-statistically-relevant results or do you see the same image I'm generating?

**Reply by briannen** [reply]
I did wind up putting the aggregated data in a different database, so based on my diff command, the databases are different between the xmls. Sorry, I forgot to share that detail!

briannen@mohawk:/d2/projects/land_nysm/metv/scorecards_troubleshoot$ diff scorecard_agg_leads.xml /d2/personal/johnhg/METplus/discussions/2734/scorecard_agg_leads.xml 
4c4
< 	<database>mv_land_nysm_stat_analysis</database>
---
> 	<database>mv_land_nysm</database>
15,17c15,17
< 	<plots>/d2/projects/land_nysm/metv/scorecards_troubleshoot/plots/</plots>
< 	<data>/d2/projects/land_nysm/metv/scorecards_troubleshoot/data/</data>
< 	<scripts>/d2/projects/land_nysm/metv/scorecards_troubleshoot/scripts/</scripts>
---
> 	<plots>/d2/personal/johnhg/METplus/discussions/2734/metv_agg/scorecards/plots/</plots>
> 	<data>/d2/personal/johnhg/METplus/discussions/2734/metv_agg/scorecards/data/</data>
> 	<scripts>/d2/personal/johnhg/METplus/discussions/2734/metv_agg/scorecards/scripts/</scripts>
141c141
<             <title>Test Aggregated Scorecard hr1.sub vs hr2.sub</title>
---
>             <title>Test Scorecard hr1.sub vs hr2.sub</title>

**Reply by JohnHalleyGotway** [reply]
@briannen, great, I'm now getting the same empty output when pulling data from this database:
```
<database>mv_land_nysm_stat_analysis</database>
```
With these log messages:
```
WARNING: event equalization removed 10 rows

INFO: Event equalization has produced no results.  Data frame is empty.
```
Using METviewer to explore the data in the `mv_land_nysm_stat_analysis` database, I do so a reason why your scorecard XML would return no data. Take a look at the `FCST_LEV` value in this screenshot.
<img width="1096" alt="Screenshot 2024-10-22 at 1 22 26 PM" src="https://github.com/user-attachments/assets/334302b8-25b3-4402-b82d-ea9dcceafb35">
It's `R5,R7,R9` meaning that's the value of the FCST_LEV column of the data that you loaded into this database.

Looking in `scorecard_agg_leads.xml`, I see you're requesting level values of `R5`, `R7`, and `R9` separately, but not `R5,R7,R9` as a group. This points to an issue with the `stat_analysis` commands you used to aggregate the data. When multiple values for a header column like `FCST_LEV` are encountered in the input, it prints a log message alerting you to the fact that it's aggregating across multiple level values and then just writes a comma-separated list of values like this.

If you want to keep output for those levels separate, just include `-by FCST_LEV` in your STAT-Analysis jobs. I'm not sure that'll solve all of the issues you're encountering, but it's definitely a change that's needed.

I'd recommend looking through the output generated by STAT-Analysis, either directly or through METviewer to confirm that all the filtering criteria you're using to define the scorecard are actually present in the data.

Hope that helps.

**Reply by JohnHalleyGotway** [reply]
@briannen I'm a bit surprised that METviewer does not already support grouping data by lead time directly, without first having to manually aggregate the data in Stat-Analysis.

I went hunting around and found [this comment](https://github.com/dtcenter/METviewer/issues/16#issuecomment-506485210) from @TatianaBurek for issue dtcenter/METviewer#16 from back in 2019 about this feature. Based on that comment, I did try grouping lead times like this in the scorecard XML:
```
        <columns>
            <field name="vx_mask">
                <val name="NY" />
                <field equalize="true" name="fcst_lead">
                    <val name="120000:240000" label="Day 1" />
                    <val name="360000:480000" label="Day 2" />
                </field>
            </field>
        </columns>
```
But that produces the following type of error messages:
```
NameError: Derived curve can't be calculated. Multiple values for one valid date/fcst_lead
```
So that definitely does not work. I'm wondering if that feature DID exist in the old R-based version of METviewer but didn't survive the transition to Python?

Regardless, I certainly do see the ability to group the input data during scorecard creation as an incredibly useful feature. @bikegeek I'd recommend writing this up as a feature request since asking users to aggregate the data themselves ahead of time is a very high bar and very error prone.

**Reply by briannen** [reply]
I see that the FCST_LEV in the stat-analysis output for TSOIL and SOILW isn't what I expect. Here is a screenshot of my METplus conf file (casper:/glade/work/briannen/metplus_land/workflow/nysm_stat_analysis.conf) where I thought I was setting the levels to loop. Is that not correct?
<img width="813" alt="Screenshot 2024-10-23 at 9 15 42 AM" src="https://github.com/user-attachments/assets/16e72f0e-ebee-444a-b100-c66d6e621787">

**Reply by georgemccabe** [reply]
Hi @briannen , as far as I can tell, your config is set up correctly to loop over each value in the `FCST_LEVEL_LIST`. Could you rerun the config setting `LOG_LEVEL=DEBUG` and send me the path to the log output? You can set this on the command line call to `run_metplus.py` by adding `config.LOG_LEVEL=DEBUG`.

**Reply by briannen** [reply]
@georgemccabe I'm running on casper, and the path is /glade/work/briannen/metplus_land/workflow/output/logs/metplus.log.20241023160319

**Reply by JohnHalleyGotway** [reply]
@briannen and @georgemccabe, I found the problem.

Looking in `metplus.log.20241023160319`, I definitely see that METplus is setting those forecast levels separately:
```
grep fcst_lev metplus.log.20241023160319 | egrep -v export | cut -d')' -f2 | sort -u
 DEBUG: Building out_stat filename from template: out_stat/{fcst_var}_{fcst_level}_{fcst_lead}_out_stat_SL1L2.stat
 DEBUG: METPLUS_FCST_LEVEL=fcst_level = ["R5"];
 DEBUG: METPLUS_FCST_LEVEL=fcst_level = ["R7"];
 DEBUG: METPLUS_FCST_LEVEL=fcst_level = ["R9"];
```
The trouble is that the STAT-Analysis config options are named `fcst_lev` and `obs_lev` on [this line](https://github.com/dtcenter/MET/blob/c5a2155ada4773b2d3c7f50d5dea8f1dff791153/data/config/STATAnalysisConfig_default#L48) of STATAnalysisConfig_default, not `fcst_level` and `obs_level` that METplus is setting.

So METplus is setting those variables, and STAT-Analysis is happily ignoring them and using the default empty lists for `fcst_lev` and `obs_lev`, which tells it to NOT FILTER by that column of data.

Seems like a METplus bug to me that requires a fix.

FWIW, I did some fancy grepping of that log file and found the following STAT-Analysis config options being exported:
```
fcst_lead 
fcst_level 
fcst_var 
jobs 
model 
obtype
```
To confirm that `fcst_level` is the only one that's not named as expected.

**Reply by georgemccabe** [reply]
Yikes! I created bugfix issue #2742 to address this. @briannen, will you need a METplus v5.1.1 bugfix release created to include this fix? Or is obtaining this fix via the main_v5.1 branch or the upcoming v6.0.0 release sufficient?

**Reply by briannen** [reply]
@georgemccabe It can wait until the v6.0.0 release.

**Comment by bikegeek** [comment]
Issue created in METviewer:
https://github.com/dtcenter/METviewer/issues/552

**Comment by jprestop** [comment]
Closing this discuss since an issue was created for this work. [dtcenter/METviewer#552](https://github.com/dtcenter/METviewer/issues/552)
